src/beliefpropagation.jl

```diff
@@ -0,0 +1,244 @@
+    sv = s[vertex][1]
+    flatpsiv = flatpsi[vertex]
+    flatpsivO = flatpsiO[vertex]
+    push!(num_tensors_to_contract, flatpsivO)
+    push!(denom_tensors_to_contract, flatpsiv)
+  end
+
+  numerator = ITensors.contract(num_tensors_to_contract)[1]
+  denominator = ITensors.contract(denom_tensors_to_contract)[1]
+
+  return numerator / denominator
+end
+
+#given two flat networks psi and psi0, calculate the ratio of their contraction centred on the subgraph(s) containing v1 and v2. The message tensors should be formulated over psi.
+function take_2sexpec_two_networks(
+  g::NamedDimGraph,
+  psi::ITensorNetwork,
+  psiO::ITensorNetwork,
+  s::IndsNetwork,
+  mts::Dict{Pair,ITensor},
+  dg_subgraphs::DataGraph,
+  v1::Tuple,
+  v2::Tuple,
+)
+  subgraph1 = find_subgraph(v1, dg_subgraphs)
+  subgraph2 = find_subgraph(v2, dg_subgraphs)
+  num_tensors_to_contract = ITensor[]
+  denom_tensors_to_contract = ITensor[]
+
+  if (subgraph1 == subgraph2)
+    connected_subgraphs = neighbors(dg_subgraphs, subgraph1)
+    for k in connected_subgraphs
+      push!(num_tensors_to_contract, mts[k => subgraph1])
+      push!(denom_tensors_to_contract, mts[k => subgraph1])
+    end
+
+    for vertex in dg_subgraphs[subgraph1]
+      if (vertex != v1 && vertex != v2)
+        push!(num_tensors_to_contract, deepcopy(psi[vertex]))
+      else
+        push!(num_tensors_to_contract, deepcopy(psiO[vertex]))
+      end
+      push!(denom_tensors_to_contract, deepcopy(psi[vertex]))
+    end
+  else
+    connected_subgraphs1 = neighbors(dg_subgraphs, subgraph1)
+    for k in connected_subgraphs1
+      if (k != subgraph2)
+        push!(num_tensors_to_contract, mts[k => subgraph1])
+        push!(denom_tensors_to_contract, mts[k => subgraph1])
+      end
+    end
+
+    connected_subgraphs2 = neighbors(dg_subgraphs, subgraph2)
+    for k in connected_subgraphs2
+      if (k != subgraph1)
+        push!(num_tensors_to_contract, mts[k => subgraph2])
+        push!(denom_tensors_to_contract, mts[k => subgraph2])
+      end
+    end
+
+    for vertex in dg_subgraphs[subgraph1]
+      if (vertex != v1)
+        push!(num_tensors_to_contract, deepcopy(psi[vertex]))
+      else
+        push!(num_tensors_to_contract, deepcopy(psiO[vertex]))
+      end
+
+      push!(denom_tensors_to_contract, deepcopy(psi[vertex]))
+    end
+
+    for vertex in dg_subgraphs[subgraph2]
+      if (vertex != v2)
+        push!(num_tensors_to_contract, deepcopy(psi[vertex]))
+      else
+        push!(num_tensors_to_contract, deepcopy(psiO[vertex]))
+      end
+
+      push!(denom_tensors_to_contract, deepcopy(psi[vertex]))
+    end
+  end
+
+  numerator = ITensors.contract(num_tensors_to_contract)[1]
+  denominator = ITensors.contract(denom_tensors_to_contract)[1]
+  out = numerator / denominator
+
+  return out
+end
+
+#Starting with initial guess for messagetensors, monitor the convergence of an observable on a single site v (which is emedded in psiflatO)
+function iterate_single_site_expec(
```

Comment:
This could be a use case for [`Observers.jl`](https://github.com/GTorlai/Observers.jl), which allows users to pass functions that will perform measurements at each iteration.

Reply:
Okay cool, yes that would be nice to weave in observers for some of the belief propagation stuff.